.github/workflows/test_dbt_models.yaml

```diff
@@ -22,31 +34,58 @@ jobs:
         with:
           role-to-assume: ${{ secrets.AWS_IAM_ROLE_TO_ASSUME_ARN }}
 
+      - name: Restore dbt state cache
+        id: cache
+        uses: ./.github/actions/restore_dbt_cache
+        with:
+          path: ${{ env.PROJECT_DIR }}/${{ env.STATE_DIR }}
+          key: ${{ env.CACHE_KEY }}
+
       - name: Install Python requirements
         run: pip install -r scripts/requirements.run_iasworld_data_tests.txt
         working-directory: ${{ env.PROJECT_DIR }}
         shell: bash
 
       - name: Run tests
         run: |
-          python3 scripts/run_iasworld_data_tests.py \
-            --target "$TARGET" \
-            --output-dir ./qc_test_results/
+          if [[ $CACHE_HIT == 'true' ]]; then
+            # shellcheck disable=SC2086
+            python3 scripts/run_iasworld_data_tests.py \
+              --target "$TARGET" \
+              --output-dir ./qc_test_results/ \
+              $SELECT_OPTION \
+              $SKIP_ARTIFACTS_OPTION \
+              --defer --state "$STATE_DIR"
+          else
+            # shellcheck disable=SC2086
+            python3 scripts/run_iasworld_data_tests.py \
+              --target "$TARGET" \
+              --output-dir ./qc_test_results/ \
+              $SELECT_OPTION \
+              $SKIP_ARTIFACTS_OPTION
+          fi
         working-directory: ${{ env.PROJECT_DIR }}
         shell: bash
         env:
           USER: ${{ github.triggering_actor }}
           GIT_SHA: ${{ github.sha }}
           GIT_REF: ${{ github.ref_name }}
           GIT_AUTHOR: ${{ github.event.commits[0].author.name }}
+          CACHE_HIT: ${{ steps.cache.outputs.cache-hit }}
+          SELECT_OPTION: ${{ inputs.select }}
+          SKIP_ARTIFACTS_OPTION: ${{ inputs.upload_test_results && '--no-skip-artifacts' || '--skip-artifacts' }}
 
       - name: Save test results to S3
+        if: inputs.upload_test_results
         run: |
           s3_prefix="s3://ccao-data-warehouse-us-east-1/qc"
           local_prefix="qc_test_results/metadata"
           for dir in "test_run" "test_run_result" "test_run_failing_row"; do
-            echo "Copying ${dir} metadata to S3"
-            aws s3 sync "${local_prefix}/${dir}" "${s3_prefix}/${dir}"
+            dirpath="${local_prefix}/${dir}"
+            if [ -e "$dirpath" ]; then
+              echo "Copying ${dirpath} metadata to S3"
+              aws s3 sync "$dirpath" "${s3_prefix}/${dir}"
+            fi
```

Comment:
See below for a description of a bug I fixed by sometimes not outputting `test_run_failing_row`.